subsys/bluetooth/controller/ll_sw/ull_adv_internal.h

```diff
@@ -185,12 +174,9 @@ uint8_t ull_adv_sync_pdu_alloc(struct ll_adv_set *adv,
 /* helper function to set/clear common extended header format fields
  * for AUX_SYNC_IND PDU.
  */
-uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync,
-				   struct pdu_adv *ter_pdu_prev,
-				   struct pdu_adv *ter_pdu,
-				   uint16_t hdr_add_fields,
-				   uint16_t hdr_rem_fields,
-				   struct ull_adv_ext_hdr_data *hdr_data);
+uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync, struct pdu_adv *ter_pdu_prev,
+				   struct pdu_adv *ter_pdu, uint16_t hdr_add_fields,
+				   uint16_t hdr_rem_fields, void *hdr_data);
```

Comment:
It actually seems to me that the last 3 arguments (`hdr_add_fields`, `hdr_rem_fields` and `hdr_data`) could benefit from being a struct, or at least `hdr_add_fields` and `hdr_data` as, if I understand correct, `hdr_data`, is the header data that will be added based on the value of `hdr_add_fields`. 

Furthermore the `void *hdr_data` value could possibly be a struct in itself, as an array of unions or something. Clearly each `ULL_ADV_PDU_HDR_FIELD_` bit defines a structured value (e.g. `ULL_ADV_PDU_HDR_FIELD_CTE_INFO` which is just a uint8_t value and `ULL_ADV_PDU_HDR_FIELD_SYNC_INFO` which is more). 

Generally the whole way this is being used is fairly hard to read as code and generally assumes that the reader is intensely familiar with all the values and implicit structs, but I suppose that has been done to try to make it as efficient as posisble?

Reply:
>It actually seems to me that the last 3 arguments (hdr_add_fields, hdr_rem_fields and hdr_data) could benefit from being a struct, or at least hdr_add_fields and hdr_data as, if I understand correct, hdr_data, is the header data that will be added based on the value of hdr_add_fields.

Yes, it could be changed into `struct`, that will add some code in calls like this: 
 https://github.com/zephyrproject-rtos/zephyr/blob/b8aa5b2ecb0410a64044f8119ea220d9d85b4096/subsys/bluetooth/controller/ll_sw/ull_df.c#L313-L315

But that is something we could live with.

> Furthermore the void *hdr_data value could possibly be a struct in itself, as an array of unions or something. Clearly each ULL_ADV_PDU_HDR_FIELD_ bit defines a structured value (e.g. ULL_ADV_PDU_HDR_FIELD_CTE_INFO which is just a uint8_t value and ULL_ADV_PDU_HDR_FIELD_SYNC_INFO which is more).

This is something I'm supporting and I should have thought about it when adding the change to the PR. 
I'm not sure if something as compact as `union` of `structs` is possible here. Pay attention that the goal of the functionality is to allow to provide content of multiple fields at once. 
But switch from compact array into `struct` would increase readability of the code.
There will be small increase of stack usage if we switch to structure holding data as members.

@cvinayak I'd like to hear your opinion about it?

Reply:
The bit fields and the value array was inspired from how Bluetooth PDU itself is formatted using the header `flag` and then the field values themself thereafter. The interface should permit multiple fields of variable lengths to be populated in a single run of the traversal of the PDU extended header format. 

Passing constant values as parameters reduces indirect memory access which is otherwise required both in the caller and callee when using `structs`.

`union` would restrict passing multiple fields, and `struct` would need worst case allocations, hence the current design of encoding the fields inspired by the extended common header format, with additional optimization to be able to pass multiple fields by reference.

This is one place multiple fields (AdvA, TgtA, ADI, AuxPtr, and TxPower flags) should be passed: https://github.com/zephyrproject-rtos/zephyr/blob/53cd7f7152da476ccfb97ce6171b43c4b195ef99/subsys/bluetooth/controller/ll_sw/ull_adv.c#L532
The code above it has not yet been ported to use `ull_adv_aux_hdr_set_clear`, which is similar to `ull_adv_sync_pdu_set_clear`

Reply:
I'm fine with `hdr_add_fields` and  `hdr_rem_fields` parameters. But in my humble opinion:
```
struct ull_adv_ext_header_data {
    struct pdu_cte_info *cte_info;
    struct {
        uint8_t len;
        uint8_t *data;
    } adv_data ;
    struct {
        uint8_t len;
        uint8_t *ret_data_ptr;
    } acad;
};
```

This would make understanding of the code far easier and lower risk of making a mistake.
The cost is a few bytes on stack + few more instructions due to indirect access. 
@cvinayak @Thalley any further opinions about it?

Reply:
Ok then. Lest make it a policy, at least for controller code,: **"Favour low RAM and ROM usage sometimes sacifying readability."**

@Thalley in such case the code stays in current form.
I'll do rebase then it will be ready for merge.

Reply:
Agreed